ensime_shared/protocol_handler.py

```diff
@@ -0,0 +1,249 @@
+import webbrowser
+import json
+
+from ensime_shared.config import gconfig, feedback, commands
+from ensime_shared.util import catch
+from ensime_shared.symbol_format import completion_to_suggest
+
+class ProtocolHandler(object):
+
+    def __init__(self):
+        self.handlers = {}
+        self.register_responses_handlers()
+
+    def register_responses_handlers(self):
+        """Register handlers for responses from the server.
+
+        A handler must accept only one parameter: `payload`.
+        """
+        self.handlers["SymbolInfo"] = self.handle_symbol_info
+        self.handlers["IndexerReadyEvent"] = self.handle_indexer_ready
+        self.handlers["AnalyzerReadyEvent"] = self.handle_analyzer_ready
+        self.handlers["NewScalaNotesEvent"] = self.buffer_typechecks
+        self.handlers["BasicTypeInfo"] = self.show_type
+        self.handlers["ArrowTypeInfo"] = self.show_type
+        self.handlers["FullTypeCheckCompleteEvent"] = self.handle_typecheck_complete
+        self.handlers["StringResponse"] = self.handle_string_response
+        self.handlers["CompletionInfoList"] = self.handle_completion_info_list
+        self.handlers["TypeInspectInfo"] = self.handle_type_inspect
+        self.handlers["SymbolSearchResults"] = self.handle_symbol_search
+        self.handlers["DebugOutputEvent"] = self.handle_debug_output
+        self.handlers["DebugBreakEvent"] = self.handle_debug_break
+        self.handlers["DebugBacktrace"] = self.handle_debug_backtrace
+        self.handlers["DebugVmError"] = self.handle_debug_vm_error
+        self.handlers["RefactorDiffEffect"] = self.apply_refactor
+        self.handlers["ImportSuggestions"] = self.handle_import_suggestions
+        self.handlers["PackageInfo"] = self.handle_package_info
+
+    def handle_incoming_response(self, call_id, payload):
+        """Get a registered handler for a given response and execute it."""
+        self.log("handle_incoming_response: in {}".format(payload))
+        typehint = payload["typehint"]
+        handler = self.handlers.get(typehint)
+        def feature_not_supported(m):
+            msg = feedback["handler_not_implemented"]
+            self.raw_message(msg.format(typehint, self.launcher.ensime_version))
+        if handler:
+            with catch(NotImplementedError, feature_not_supported):
+                handler(call_id, payload)
```

Comment:
I disagree. I find very useful to split error handling logic into different local functions, even if they cannot be immediately reusable. IMO, I find it the right way to do it from a modularity standpoint (and following fp good practices). I understand this is not idiomatic of Python, but I see several benefits of wrapping everything into `catch`:

- It's damn easy to add more error handling statements without polluting what a function really does. Usually, when browsing code, we don't care about what happens if there's a failure, especially if we're learning a codebase. We just want to know how the logic proceeds. It's very annoying when I have to browse a codebase and I find 25-lines in the `catch` block that prevents me from focusing on the real important parts.
- The error handling logic can be reused in most of the times. This is not true for very concrete/idiomatic cases, for instance when catching `KeyError` or `NotImplemented`, but I consider healthier to go for other alternatives. Concretely, if a value is not in a dictionary, I prefer `get` instead of just expecting an exception to happen.
- One can use `catch` not to only execute handlers, but enrich the try-catch functionality with logging *without anyone noticing*. It is not only useful for debugging purposes, but also to check when some weird conditions happen and they shouldn't.
  
I don't know how people are using `catch`, but I would use it only to wrap statements that produce an exception. In fact, in case that more than one happens, it should wrap the minimum possible lines that cover all of them.